lingua_franca/Formatter.py

```diff
@@ -0,0 +1,593 @@
+from lingua_franca.common import \
+                populate_localized_function_dict, _localized_function_caller,\
+                    _SUPPORTED_LANGUAGES
+from lingua_franca.lang import get_primary_lang_code, get_full_lang_code
+
+from lingua_franca.bracket_expansion import SentenceTreeParser
+
+from collections import namedtuple
+from inspect import signature
+from warnings import warn
+from os.path import join
+import json
+import os
+import datetime
+import re
+
+_REGISTERED_FUNCTIONS = ["nice_number",
+                         "nice_time",
+                         "pronounce_number",
+                         "nice_response",
+                         "nice_ordinal",
+                         "nice_part_of_day"]
+
+NUMBER_TUPLE = namedtuple(
+    'number',
+    ('x, xx, x0, x_in_x0, xxx, x00, x_in_x00, xx00, xx_in_xx00, x000, ' +
+    'x_in_x000, x0_in_x000, x_in_0x00'))
+
+class Formatter:
+    ''' An object-oriented interface to Lingua Franca. Can load only those
+        which are currently in use, potentially saving on memory. By default,
+        it will load all languages when instantiated.
+
+        Arguments:
+          langs (str or list(str)): the language code or languages codes which
+                                    you would like to initialize
+
+        Use just like the `format` module:
+            >>>from lingua_franca.Formatter import Formatter
+            >>>fmt = Formatter()
+            >>>fmt.pronounce_number(13)
+            'thirteen'
+            >>>fmt.pronounce_number(13, lang="es")
+            'trece'
+
+        Or initialize it with only the languages you need:
+            >>>fmt = Formatter(["en", "es"])
+
+        It can also accept a single language code as a string, and compensate
+        for the fact that English is missing:
+            >>>fmt = Formatter("es"):
+            >>>fmt.pronounce_number(13)
+            'trece'
+
+
+    '''
+    _LOCALIZED_FUNCTIONS = {}
+    default_language = ""
```

Comment:
this should be replaced with ```lingua_franca.lang.get_active_lang```

Reply:
I was wondering that. Should it, or should we dispose of the other? `lingua_franca.lang.get_active_lang` doesn't appear to be called anywhere at the moment. This would be its first dance outside of some tests.

Reply:
i dont want to have to remember downstream to always pass a lang parameter down the pipeline. a global config is great, keeps language selection in a logical place instead of being something usually constant you keep passing along (every func call / formatter or parser object initialization )

Other way to look at it is that without it we are forcing english as a default by no particular reason

Reply:
Right. How about moving it to `common.py`, to live with `_SUPPORTED_LANGUAGES`?

Reply:
i dont think so. since this facing end users consuming the library, id rather keep it in `lingua_franca.lang`  id like to keep common.py as a "private" thing internal to the library not meant to be imported

Reply:
Tell you the truth, I'd rather end users didn't need to import either of those. The *only* things still in `lang/init.py` are the lang code, and active_lang getters and setters.

If an end user does `from lingua_franca import lang` instead of `from lingua_franca.lang import get_primary_lang_code, set_primary_lang_code`, they're completely obviating any memory gains from the refactor.

Similarly, moving those things up to `lingua_franca/init.py` would invariably result in people importing the module wholesale, just to get at those functions.

So, *if* one of those files has to be public-facing, I think `common.py` is probably safest. On the other hand, `resolve_resource_file()` is the only current member of `lingua_franca.common` that users might call, and it could move, or the rest of `common.py` could move to something like `function_mgmt.py`

Reply:
Alright, having given it some thought, how about:

1. Move actual definitions from `lingua_franca.lang` to `lingua_franca.common`, such as `get_active_lang()`
2. Replace them with passthroughs, slap a `PendingDeprecationWarning` on them, replace with a `DeprecationWarning` in v0.3.0, and remove them at some future date
3. Move non-user facing functions *from* `lingua_franca.common` *to* `lingua_franca.lang`

This would:

* allow us to make it explicit that users should not, generally speaking, import `lingua_franca.lang` or any of its members
* place those functions in a module that will make sense to users: the active languages and language codes are `common` to all the other modules in the package
* implicitly discourage `import lingua_franca`

Reply:
That idea also re-raises the notion (I think you're the one who suggested it) of removing language codes from localized functions, so that somebody *could* do, `from lingua_franca.lang import format_en as format` and get the same result as creating a `Formatter("en")`.

On the other hand, collisions are *impossible* when the localized functions all have different names.